File: highway-pursuit-server/highway-pursuit-server/Injected/InputService.cs
using HighwayPursuitServer.Data;
using HighwayPursuitServer.Server;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;

namespace HighwayPursuitServer.Injected
{
    class InputService
    {
        private readonly Dictionary<Input, uint> _inputToOffset = new Dictionary<Input, uint>{
            {Input.Accelerate, MemoryAdresses.ACCELERATE_OFFSET},
            {Input.Brake, MemoryAdresses.BRAKE_OFFSET},
            {Input.SteerL, MemoryAdresses.STEER_L_OFFSET},
            {Input.SteerR, MemoryAdresses.STEER_R_OFFSET},
            {Input.Fire, MemoryAdresses.FIRE_OFFSET},
            {Input.Oil, MemoryAdresses.OIL_OFFSET},
            {Input.Smoke, MemoryAdresses.SMOKE_OFFSET},
            {Input.Missiles, MemoryAdresses.MISSILES_OFFSET},
        };

        private List<Input> _currentInputs = new List<Input>();
        private const byte ACTIVE_KEY = 0x80;

#if DEBUG
        private const byte MANUAL_CONTROL_KEY = 0x2A; // holding left shift enables manual keyboard control
#endif

        private readonly IHookManager _hookManager;

        public InputService(IHookManager hookManager)
        {
            this._hookManager = hookManager;
            this.RegisterHooks();
        }

        #region input management
        public int GetInputCount()
        {
            return _inputToOffset.Count;
        }
        
        public void SetInput(List<Input> inputs)
        {
            this._currentInputs = inputs;
        }

        private int InputToKeyCode(Input input)
        {
            var offset = _inputToOffset[input];
            var module = _hookManager.GetModuleBase();
            return Marshal.ReadInt32(new IntPtr(module.ToInt32() + offset));
        }
        #endregion

        #region Hooking
        private void RegisterHooks()
        {
            IntPtr setCoopLevelPtr = new IntPtr(_hookManager.GetDINPUTBase().ToInt32() + MemoryAdresses.SET_COOPERATIVE_LEVEL_OFFSET);
            SetCooperativeLevel = Marshal.GetDelegateForFunctionPointer<SetCooperativeLevel_delegate>(setCoopLevelPtr);
            _hookManager.RegisterHook(setCoopLevelPtr, new SetCooperativeLevel_delegate(SetCooperativeLevel_Hook));

            IntPtr getDeviceStatePtr = new IntPtr(_hookManager.GetDINPUTBase().ToInt32() + MemoryAdresses.GET_DEVICE_STATE_OFFSET);
            GetDeviceState = Marshal.GetDelegateForFunctionPointer<GetDeviceState_delegate>(getDeviceStatePtr);
            _hookManager.RegisterHook(getDeviceStatePtr, new GetDeviceState_delegate(GetDeviceState_Hook));
        }

        #region hooks
        uint SetCooperativeLevel_Hook(IntPtr pInput, IntPtr hwnd, DISCL_FLAGS flags)
        {
            // Make it so the app doesn't require foreground or exclusive access, to avoid failing keyboard acquisition
            DISCL_FLAGS nonBlockingFlags = DISCL_FLAGS.DISCL_BACKGROUND | DISCL_FLAGS.DISCL_NONEXCLUSIVE;
            return SetCooperativeLevel(pInput, hwnd, nonBlockingFlags);
        }

        uint GetDeviceState_Hook(IntPtr pInput, uint deviceSize, IntPtr pDeviceStateArray)
        {
            uint res = GetDeviceState(pInput, deviceSize, pDeviceStateArray);
#if DEBUG
            // Don't do anything in manual control
            byte[] originalState = new byte[deviceSize];
            Marshal.Copy(pDeviceStateArray, originalState, 0, (int)deviceSize);
            if (originalState[MANUAL_CONTROL_KEY] == ACTIVE_KEY)
            {
                return res;
            }
#endif

            // Edit the keyboard state
            byte[] modifiedState = new byte[deviceSize];

            foreach (Input input in _currentInputs)
            {
                var keycode = InputToKeyCode(input);
                modifiedState[keycode] = ACTIVE_KEY;
            }

            Marshal.Copy(modifiedState, 0, pDeviceStateArray, (int)deviceSize);
            return res;
        }
        #endregion
        #region original function pointers
        static SetCooperativeLevel_delegate SetCooperativeLevel;
        static GetDeviceState_delegate GetDeviceState;
        #endregion

        #region delegates
        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
        [return: MarshalAs(UnmanagedType.U4)]
        delegate uint SetCooperativeLevel_delegate(IntPtr pInput, IntPtr hwnd, DISCL_FLAGS flags);

        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
        [return: MarshalAs(UnmanagedType.U4)]
        delegate uint GetDeviceState_delegate(IntPtr pInput, uint cbData, IntPtr lpvData);
        #endregion
        #endregion
    }
}

File: highway-pursuit-server/highway-pursuit-server/Injected/InputService.cpp
#include "../pch.h"
#include "InputService.hpp"
#include "MemoryAddresses.hpp"

namespace Injected
{
    InputService::InputService(std::shared_ptr<HookManager> hookManager)
        : _hookManager(hookManager)
    {
        FindAddresses();
        RegisterHooks();
    }

    int InputService::GetInputCount()
    {
        return HighwayPursuitConstants::ACTION_COUNT;
    }

    void InputService::SetInput(const std::vector<Input>& inputs)
    {
        _currentInputs = inputs;
    }

    uint32_t InputService::InputToOffset(Input input)
    {
        switch (input)
        {
        case Input::Accelerate:
            return MemoryAddresses::ACCELERATE_OFFSET;
        case Input::Brake:
            return MemoryAddresses::BRAKE_OFFSET;
        case Input::SteerL:
            return MemoryAddresses::STEER_L_OFFSET;
        case Input::SteerR:
            return MemoryAddresses::STEER_R_OFFSET;
        case Input::Fire:
            return MemoryAddresses::FIRE_OFFSET;
        case Input::Oil:
            return MemoryAddresses::OIL_OFFSET;
        case Input::Smoke:
            return MemoryAddresses::SMOKE_OFFSET;
        case Input::Missiles:
            return MemoryAddresses::MISSILES_OFFSET;
        default:
            throw HighwayPursuitException(ErrorCode::UNKNOWN_ACTION);
        }
    }

    // InputToKeyCode method
    uint32_t InputService::InputToKeyCode(Input input)
    {
        auto offset = InputToOffset(input);
        auto module = _hookManager->GetModuleBase();
        uint32_t* keyCode = reinterpret_cast<uint32_t*>(module + offset);
        return *keyCode;
    }

    void InputService::FindAddresses()
    {
        // Get entry point
        DirectInputCreate_Base = reinterpret_cast<DirectInput8Create_t>(GetProcAddress(_hookManager->GetDirectInput8(), "DirectInput8Create"));
        if (DirectInputCreate_Base == nullptr)
        {
            throw std::runtime_error("Failed to get DirectInputCreate");
        }

        // Create direct input
        HMODULE hInstance = GetModuleHandle(NULL);
        DirectInput8Wrapper dinput(hInstance, HighwayPursuitConstants::DINPUT_VERSION, HighwayPursuitConstants::IID_IDirectInput8A);
        uintptr_t* dinput_vtable = *reinterpret_cast<uintptr_t**>(dinput.DirectInput());

        DirectInputRelease_Base = reinterpret_cast<DirectInput8Release_t>(dinput_vtable[MemoryAddresses::DINPUT_RELEASE_OFFSET]);
        DirectInputCreateDevice_Base = reinterpret_cast<DirectInput8CreateDevice_t>(dinput_vtable[MemoryAddresses::DINPUT_CREATE_DEVICE_OFFSET]);

        // Create direct input device
        DirectInputDevice8Wrapper device(dinput.DirectInput(), HighwayPursuitConstants::GUID_SysKeyboard);
        uintptr_t* device_vtable = *reinterpret_cast<uintptr_t**>(device.Device());

        DirectInputReleaseDevice_Base = reinterpret_cast<DirectInput8ReleaseDevice_t>(device_vtable[MemoryAddresses::DINPUT_RELEASE_DEVICE_OFFSET]);
        SetCooperativeLevel_Base = reinterpret_cast<SetCooperativeLevel_t>(device_vtable[MemoryAddresses::SET_COOPERATIVE_LEVEL_OFFSET]);
        GetDeviceState_Base = reinterpret_cast<GetDeviceState_t>(device_vtable[MemoryAddresses::GET_DEVICE_STATE_OFFSET]);
    }

    // RegisterHooks method
    void InputService::RegisterHooks()
    {
        InputService::Instance = this;

        _hookManager->RegisterHook(SetCooperativeLevel_Base, &SetCooperativeLevel_StaticHook, &SetCooperativeLevel_Base);
        _hookManager->RegisterHook(GetDeviceState_Base, &GetDeviceState_StaticHook, &GetDeviceState_Base);
    }

    void InputService::HandleD3DERR(D3DERR errorCode)
    {
        if (errorCode > 0)
        {
            throw D3D8Exception(errorCode);
        }
    }

    D3DERR InputService::SetCooperativeLevel_Hook(IDirectInputDevice8* pInput, HWND hwnd, DISCL_FLAGS dwFlags)
    {
        // Make it so the app doesn't require foreground or exclusive access, to avoid failing keyboard acquisition
        DISCL_FLAGS nonBlockingFlags = DISCL_FLAGS::DISCL_BACKGROUND | DISCL_FLAGS::DISCL_NONEXCLUSIVE;
        return SetCooperativeLevel_Base(pInput, hwnd, nonBlockingFlags);
    }

    D3DERR InputService::GetDeviceState_Hook(IDirectInputDevice8* pInput, DWORD cbData, LPVOID lpvData)
    {
        D3DERR res = GetDeviceState_Base(pInput, cbData, lpvData);
        uint32_t deviceSize = cbData;
        uint8_t* pDeviceState = reinterpret_cast<uint8_t*>(lpvData);

        // Don't do anything in manual control
        if (pDeviceState[MANUAL_CONTROL_KEY] == ACTIVE_KEY)
        {
            return res;
        }

        // Update the modifiedState based on the inputs
        for (Input input : _currentInputs)
        {
            int keycode = InputToKeyCode(input);
            pDeviceState[keycode] = ACTIVE_KEY;
        }
        return res;
    }

    D3DERR __stdcall InputService::SetCooperativeLevel_StaticHook(IDirectInputDevice8* pInput, HWND hwnd, DISCL_FLAGS dwFlags)
    {
        if (InputService::Instance != nullptr)
        {
            return InputService::Instance->SetCooperativeLevel_Hook(pInput, hwnd, dwFlags);
        }
        return D3DERR_NOTAVAILABLE;
    }
    
    D3DERR __stdcall InputService::GetDeviceState_StaticHook(IDirectInputDevice8* pInput, DWORD cbData, LPVOID lpvData)
    {
        if (InputService::Instance != nullptr)
        {
            return InputService::Instance->GetDeviceState_Hook(pInput, cbData, lpvData);
        }
        return D3DERR_NOTAVAILABLE;
    }


    InputService::DirectInput8Wrapper::DirectInput8Wrapper(HMODULE hinst, DWORD dwVersion, REFIID riidltf)
    {
        IDirectInput8* dinput;
        HandleD3DERR(DirectInputCreate_Base(hinst, dwVersion, riidltf, &dinput, nullptr));
        _dinput = dinput;
    }

    InputService::DirectInput8Wrapper::~DirectInput8Wrapper()
    {
        HandleD3DERR(DirectInputRelease_Base(_dinput));
    }

    IDirect3D8* InputService::DirectInput8Wrapper::DirectInput() const
    {
        return _dinput;
    }

    InputService::DirectInputDevice8Wrapper::DirectInputDevice8Wrapper(IDirectInput8* dinput, REFGUID rguid)
    {
        IDirectInputDevice8* pDevice;
        HandleD3DERR(DirectInputCreateDevice_Base(dinput, rguid, &pDevice, nullptr));
        _pDevice = pDevice;
    }

    InputService::DirectInputDevice8Wrapper::~DirectInputDevice8Wrapper()
    {
        HandleD3DERR(DirectInputReleaseDevice_Base(_pDevice));
    }

    IDirect3DSurface8* InputService::DirectInputDevice8Wrapper::Device() const
    {
        return _pDevice;
    }

    InputService* InputService::Instance = nullptr;
    InputService::DirectInput8Create_t InputService::DirectInputCreate_Base = nullptr;
    InputService::DirectInput8Release_t InputService::DirectInputRelease_Base = nullptr;
    InputService::DirectInput8CreateDevice_t InputService::DirectInputCreateDevice_Base = nullptr;
    InputService::DirectInput8ReleaseDevice_t InputService::DirectInputReleaseDevice_Base = nullptr;
    InputService::SetCooperativeLevel_t InputService::SetCooperativeLevel_Base = nullptr;
    InputService::GetDeviceState_t InputService::GetDeviceState_Base = nullptr;
}